### ISSUE TITLE
Java Path error in windows

### ISSUE THREAD

**Issue by shrjoshi** [prashant-ramcharan/courgette-jvm#173]
I am try to execute feature in parallel using courgette-jvm

- Java Version: 8
- OS: windows 10
- Courgette-jvm version:4.3.4
- Gradle dependency:
        testCompile group: 'io.github.prashant-ramcharan', name: 'courgette-jvm', version: '4.3.4'

**Runner config:**

@RunWith(Courgette.class)
@CourgetteOptions(
        threads = 5,
        runLevel = CourgetteRunLevel.FEATURE,
        rerunFailedScenarios = false,
//        rerunAttempts = 1,
        showTestOutput = true,
        reportTargetDir = "output",
        shortenJavaClassPath=true,
        cucumberOptions = @CucumberOptions(
                features = "src/test/resources/com.testing.features",
                glue = {"com.test.gluecode"},
                tags = {"@TestParallel"},
                plugin = {
                        "pretty",
                        "json:build/cucumber-report/cucumber.json",
                        "html:build/cucumber-report/cucumber.html",
                        "junit:build/cucumber-report/cucumber.xml"},
                strict = true

        ))

public class TestParallelRunner {

}

When I execute the runner I get below error:
java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/Program%20Files%20(x86)/JetBrains/IntelliJ%20IDEA%202018.2.2/lib/idea_rt.jar
	at sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
	at sun.nio.fs.WindowsPath.parse(WindowsPath.java:94)
	at sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:255)
	at java.nio.file.Paths.get(Paths.java:84)
	at courgette.runtime.utils.FileUtils.lambda$copyClassPathFilesToTempDirectory$1(FileUtils.java:107)
	at java.util.Arrays$ArrayList.forEach(Arrays.java:3880)
	at courgette.runtime.utils.FileUtils.copyClassPathFilesToTempDirectory(FileUtils.java:105)
	at courgette.runtime.CourgetteRunner.run(CourgetteRunner.java:51)
	at courgette.api.junit.Courgette.run(Courgette.java:48)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

Process finished with exit code -2

**On further debugging, I found that the piece of code that is causing this issues is:**
_FileUtils.java
 Path source = Paths.get(classPathUrl.getPath());
                Path target = Paths.get(tmpDir.toFile().getAbsoluteFile() + File.separator + source.getFileName());
                Files.copy(source, target, StandardCopyOption.REPLACE_EXISTING);_

The class is **FINAL** so I can not override the behavior.

I can not upgrade to java 9/10.

Is there a solution to this ? 
It works well on non windows machine but it takes way to much time as I can not run code locally.
Any help is appreciated.

**Comment by prashant-ramcharan** [comment]
This looks like an IntelliJ issue. 

If you run the tests using gradle or maven this is unlikely to happen.

Can you please try that?

**Comment by prashant-ramcharan** [comment]
Also, if your classpath does not exceed the limit then you don’t need to include this option in your courgette runner:

shortenJavaClassPath

**Comment by shrjoshi** [comment]
Thanks for the quick response

- shortenJavaClassPath=false is not a option

**Now i ran the same from command line:**

$ ./gradlew runFeaturesUsingJUnit
> Task :compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava UP-TO-DATE
> Task :processTestResources UP-TO-DATE
> Task :testClasses UP-TO-DATE

> Task :runFeaturesUsingJUnit

com.testing.features > initializationError FAILED
    java.nio.file.InvalidPathException

1 test completed, 1 failed

> Task :runFeaturesUsingJUnit FAILED

FAILURE: Build failed with an exception.

**Comment by prashant-ramcharan** [comment]
Ok, I will investigate further.

**Comment by prashant-ramcharan** [comment]
I think the issue is caused by the / in the file path on Windows OS.

I will release a fix for this issue.

**Comment by prashant-ramcharan** [comment]
@shrjoshi can you please try with the following version and let me know if it solves your issue?

[4.4.1.zip](https://github.com/prashant-ramcharan/courgette-jvm/files/4458898/4.4.1.zip)

**Comment by shrjoshi** [comment]
Does the Jar have correct dependencies bundled up ? using the zip version you gave me gives lot of issues with cucumber transitive dependencies.
few mentions
com.google.gson.JsonObject;
com.google.gson.Gson;
etc

However, the published version that i was using earlier does not give this problem.
Can you please check once ? Meantime I will try to see if i can fix the dependencies.

**Comment by prashant-ramcharan** [comment]
Unzip the folder to your maven directory `C:\Users\{your-username}\.m2\repository\io\github\prashant-ramcharan\courgette-jvm`

In your build.gradle add this

````
repositories {
    mavenLocal()
}

dependencies {
  testCompile 'io.github.prashant-ramcharan:courgette-jvm:4.4.1'
}
````

**Comment by prashant-ramcharan** [comment]
If you still have issues verifying.. I can setup a Windows VM tomorrow to verify.

**Comment by prashant-ramcharan** [comment]
@shrjoshi I've now released version 4.4.1 which fixes this issue.

**Comment by shrjoshi** [comment]
Thanks for releasing the new version.
The new version does not give me the path error but even if I pass the cucumber option ‘shortenJavaClassPath=true’, still getting error as flie extension too long. The flag is in effective. Thanks for your support

**Comment by shrjoshi** [comment]
You have closed the issue.So should I assume that it will not work on windows ?

**Comment by prashant-ramcharan** [comment]
Courgette does work on Windows but not when your project uses file names that exceed the classpath length. 

**Windows Create Process has 32K limitation on the command line**

If you google `CreateProcess error=206` you will find a lot of libraries that have the same issue. This is not an issue in the library but more of a limitation on Windows when using the URL class loader.

You could try some of the suggestions noted here:

https://support.thoughtworks.com/hc/en-us/articles/213248526-Getting-around-maximum-command-line-length-is-32767-characters-on-Windows

https://stackoverflow.com/a/10598074

**Comment by prashant-ramcharan** [comment]
@shrjoshi please use Courgette version `4.5.0`. I've changed the Courgette runner to omit the creation of the classpath at runtime. This should resolve the  `file extension too long error`. This change also means that the `shortenJavaClassPath` option is no longer needed / supported.

Let me know if this version works in your project.

**Comment by prashant-ramcharan** [comment]
**Update**: There is a workaround for this in Courgette version `5.9.2` that accepts a `custom class path`.

See example project here: https://github.com/prashant-ramcharan/courgette-jvm-classpath-example
